[qunat][fx][fix] Fix get_module_type for fusion

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #73345
* __->__ #72735
* #73344

Summary:
1. We use `get_matched_types` to get the (type) pattern from matched modules.
And we need to use MatchAllNode instead of type(MatchAllNode) to query the fuser_method for the pattern

2. For the matched pattern, previous we'll skip the nodes that is in the pattern, except for the root node, e.g. if we have conv - bn - relu, we'll skip bn and relu, and only process conv, but this is no longer correct when we have more complex patterns, especially when we have MatchAllNode in the pattern. since we need to copy the nodes that matched to MatchAllNodes. Therefore we added a `node_to_subpattern` dictionary in find_matches, so that we know the corresponding subpattern for a given node, if the node is matched with MatchAllNode, we will copy the node instead of ignoring it

3.  Another fix we have in this PR is that we fixed get_fuser_method_new to allow wildcard look up, see comments for `get_valid_patterns` in `fuser_method_mappings.py`, we can get all valid patterns for a given op pattern and try to get the fuser method for these patterns

Test Plan:
python test/test_quantization.py TestFuseFx.test_fuse_custom_pattern

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34180705](https://our.internmc.facebook.com/intern/diff/D34180705)

jerryzh168 committed:
[qunat][fx][fix] Fix get_module_type for fusion

Summary:
We use `get_matched_types` to get the (type) pattern from matched modules.
And we need to use MatchAllNode instead of type(MatchAllNode) to query the fuser_method for the pattern

Test Plan:
TODO

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34180705).

jerryzh168 committed:
Update on "[qunat][fx][fix] Fix get_module_type for fusion"


Summary:
We use `get_matched_types` to get the (type) pattern from matched modules.
And we need to use MatchAllNode instead of type(MatchAllNode) to query the fuser_method for the pattern

Test Plan:
TODO

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34180705](https://our.internmc.facebook.com/intern/diff/D34180705)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34180705).

jerryzh168 committed:
Update on "[qunat][fx][fix] Fix get_module_type for fusion"


Summary:
We use `get_matched_types` to get the (type) pattern from matched modules.
And we need to use MatchAllNode instead of type(MatchAllNode) to query the fuser_method for the pattern

Test Plan:
TODO

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34180705](https://our.internmc.facebook.com/intern/diff/D34180705)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34180705).

jerryzh168 committed:
Update on "[qunat][fx][fix] Fix get_module_type for fusion"


Summary:
We use `get_matched_types` to get the (type) pattern from matched modules.
And we need to use MatchAllNode instead of type(MatchAllNode) to query the fuser_method for the pattern

Test Plan:
TODO

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34180705](https://our.internmc.facebook.com/intern/diff/D34180705)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34180705).

jerryzh168 committed:
Update on "[qunat][fx][fix] Fix get_module_type for fusion"


Summary:
1. We use `get_matched_types` to get the (type) pattern from matched modules.
And we need to use MatchAllNode instead of type(MatchAllNode) to query the fuser_method for the pattern

2. For the matched pattern, previous we'll skip the nodes that is in the pattern, except for the root node, e.g. if we have conv - bn - relu, we'll skip bn and relu, and only process conv, but this is no longer correct when we have more complex patterns, especially when we have MatchAllNode in the pattern. since we need to copy the nodes that matched to MatchAllNodes. Therefore we added a `node_to_subpattern` dictionary in find_matches, so that we know the corresponding subpattern for a given node, if the node is matched with MatchAllNode, we will copy the node instead of ignoring it

Test Plan:
python test/test_quantization.py TestFuseFx.test_fuse_custom_pattern

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34180705](https://our.internmc.facebook.com/intern/diff/D34180705)

[ghstack-poisoned]

raghuramank10000 commented on torch/ao/quantization/fx/fusion_patterns.py:
```diff
@@ -77,38 +77,40 @@ def fuse(self,
-            return type(m)
+            if isinstance(m, torch.nn.Module):
+                return type(m)
+            return m
```
Why does get_matched_types return the module itself when it is not of type nn.module? Is this intended?

jerryzh168 replied:
yeah we do return the type when it is an instance of `torch.nn.Module`, see above. otherwise we need to return m itself, this is the cases when m is torch ops or functionals, like operator.add, torch.add, torch.nn.functional.relu etc.

raghuramank10000 reviewed (approved):
Good to go once comments are addressed.

jerryzh168 committed:
Update on "[qunat][fx][fix] Fix get_module_type for fusion"


Summary:
1. We use `get_matched_types` to get the (type) pattern from matched modules.
And we need to use MatchAllNode instead of type(MatchAllNode) to query the fuser_method for the pattern

2. For the matched pattern, previous we'll skip the nodes that is in the pattern, except for the root node, e.g. if we have conv - bn - relu, we'll skip bn and relu, and only process conv, but this is no longer correct when we have more complex patterns, especially when we have MatchAllNode in the pattern. since we need to copy the nodes that matched to MatchAllNodes. Therefore we added a `node_to_subpattern` dictionary in find_matches, so that we know the corresponding subpattern for a given node, if the node is matched with MatchAllNode, we will copy the node instead of ignoring it

Test Plan:
python test/test_quantization.py TestFuseFx.test_fuse_custom_pattern

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34180705](https://our.internmc.facebook.com/intern/diff/D34180705)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34180705).

jerryzh168 committed:
Update on "[qunat][fx][fix] Fix get_module_type for fusion"


Summary:
1. We use `get_matched_types` to get the (type) pattern from matched modules.
And we need to use MatchAllNode instead of type(MatchAllNode) to query the fuser_method for the pattern

2. For the matched pattern, previous we'll skip the nodes that is in the pattern, except for the root node, e.g. if we have conv - bn - relu, we'll skip bn and relu, and only process conv, but this is no longer correct when we have more complex patterns, especially when we have MatchAllNode in the pattern. since we need to copy the nodes that matched to MatchAllNodes. Therefore we added a `node_to_subpattern` dictionary in find_matches, so that we know the corresponding subpattern for a given node, if the node is matched with MatchAllNode, we will copy the node instead of ignoring it

3.  Another fix we have in this PR is that we fixed get_fuser_method_new to allow wildcard look up, see comments for `get_valid_patterns` in `fuser_method_mappings.py`, we can get all valid patterns for a given op pattern and try to get the fuser method for these patterns

Test Plan:
python test/test_quantization.py TestFuseFx.test_fuse_custom_pattern

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D34180705](https://our.internmc.facebook.com/intern/diff/D34180705)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34180705).

vkuzo commented on torch/ao/quantization/fuser_method_mappings.py:
```diff
@@ -226,6 +227,37 @@ def reversed(is_qat, x, w):
     (nn.BatchNorm3d, nn.ConvTranspose3d): reverse2(fuse_convtranspose_bn),
 }
 
+def get_valid_patterns(op_pattern):
```
could we explain why this is needed? seems like the only place this is used is for getting a fuser method, what's the use case for getting a fuser method where one of the nodes is a wildcard?

jerryzh168 replied:
this is because the node that is supposed to be matched with MatchAllNode could be some normal node, for example, nn.MaxPool, we have this example in the test case, we are trying to match
(torch.add, MatchAllNode, (torch.nn.ReLU, torch.nn.Conv2d)) and the actual pattern we have is:
(torch.add, torch.nn.MaxPool, (torch.nn.ReLU, torch.nn.Conv2d)) --> here torch.nn.MaxPool actually can match MatchAllNode, so we need to generate this key in fuser_method lookup (fuser_method_mapping maps from a pattern to the fuser_method corresponding to it.

Alternatively we could also change the way we do lookup I think, in that case, we need to go through all patterns in the fuser_method_mapping one by one (this makes sense as well I think)

vkuzo replied:
`(torch.add, MatchAllNode, (torch.nn.ReLU, torch.nn.Conv2d))`

is the fusion going to be implemented with wildcard support?  Some kind of backend specific fusion?  Is there some example of how this works in a real backend?  I think I'm just confused on how can a backend fuse a wildcard node without knowing the node type.

jerryzh168 replied:
we have an example in `test_fuse_custom_pattern`. 
re "how can a backend fuse a wildcard node without knowing the node type." - it's just for input I think, `(torch.add, MatchAllNode, (torch.nn.ReLU, torch.nn.Conv2d))` just means the fusion can work no matter what node the first input of add is, actually this is True for all patterns, maybe to make everything consistent, this pattern should be:

```
(torch.add, MatchAllNode, (torch.nn.ReLU, (torch.nn.Conv2d, MatchAllNode)))
```
since we don't care about the type for input of torch.nn.Conv2d either

vkuzo replied:
hmm, should we have a distinction between "match any input" versus "match any node" in the matching specification?  I think I understand the use case now, but the pattern looks like it's actually matching a node and not the input.

jerryzh168 replied:
this is something inherent in fx node.. fx node is both a Node and the output Value. "input" is a concept of Value, not Node, but we don't have such thing in FX as a separate concept right now I think.
I feel that to be consistent, this pattern is probably better than the current pattern though: 
```
(torch.add, MatchAllNode, (torch.nn.ReLU, (torch.nn.Conv2d, MatchAllNode)))
```

jerryzh168 replied:
To respond to the question directly, the difference between "match any input" and "match any node" would be the first one doesn't include the input in the pattern and the second does? I feel currently we doesn't seem to have use cases for "match any node", it's hard to imagine how a pattern could include arbitrary node, so we probably won't have that use case?

I can think of use cases like, match some elementwise ops, or match some reduction ops, but for these we can just spell out all the patterns for now I think.

vkuzo replied:
Sure, it makes sense after your comments.  Maybe we can use naming or documentation in the future to make this clearer.

vkuzo commented on torch/ao/quantization/fuser_method_mappings.py:
```diff
@@ -235,6 +267,11 @@ def get_fuser_method_new(
         fuser_method_mapping = DEFAULT_PATTERN_TO_FUSER_METHOD
 
-    fuser_method = fuser_method_mapping.get(op_pattern, None)
+    op_patterns = get_valid_patterns(op_pattern)
```
could we clarify the naming?  `op_patterns = some_function(op_pattern)` is confusing because the name of the result is only one character away from the name of the input variable

jerryzh168 replied:
yeah sure
